Switch from nomnom to commander

edemaine opened the pull request:
* Fix #1489
* ~~`--macro` repeatable option replaced with single `--macros` option with ";" separator (until https://github.com/tj/commander.js/issues/674)~~ Never mind, this feature was already added, so `--macro` is now back to normal.
* Add missing `--max-expand` option
* Replace expired `--unicode-text-in-math-mode` option with `--strict` (https://github.com/Khan/KaTeX/issues/1483#issuecomment-405108291)
* Add instructions for how to run in the opening comments
* Clean up option doc strings

ylemkimon commented:
`dockers/Screenshotter/screenshotter.js` and `test/symgroups.js` also uses `nomnom`.

ylemkimon commented on cli.js:
```diff
@@ -1,103 +1,53 @@
 /* eslint no-console:0 */
 
 const katex = require("./");
+const json = require("./package.json");
```
I think the name `package` or using object destructuring, `{version}` is better.

ylemkimon commented on cli.js:
```diff
@@ -1,103 +1,53 @@
-    .parse();
+const options = require("commander")
+    .version(json.version)
+    .option("-d, --display-mode", "displayMode",
```
The second argument seems to be passed by a mistake.

ylemkimon commented on cli.js:
```diff
@@ -1,103 +1,53 @@
+
 /* eslint no-console:0 */
 
 const katex = require("./");
```
How about wrapping this with try...catch and printing helpful message when the module is not found, i.e., KaTeX is not built?

ylemkimon commented on package-lock.json:
```diff
@@ -2053,16 +2048,6 @@
         "lazy-cache": "^1.0.3"
       }
     },
-    "chalk": {
```
Could you try to run `npm dedupe` or deduplicate `chalk` and `ansi-style`? It seems they are duplicated many times.

edemaine commented:
Ah, I see. Running `screenshotter.sh` (and thus `webpack`) from the `dockers/Screenshotter` directory fails, while running `dockers/Screenshotter/screenshotter.sh` from the root directory succeeds.  Is this a bug in https://github.com/gnandretta/babel-plugin-version-inline/blob/master/src/index.js#L3?  I'm not sure how paths are supposed to work in webpack.  @ylemkimon?

ylemkimon commented on dockers/Screenshotter/screenshotter.js:
```diff
@@ -109,11 +67,13 @@ if (opts.exclude) {
+const numAttempts = opts.attempts || 5;
+let seleniumURL = opts.seleniumUrl;
+let seleniumIP = opts.seleniumIp;
+let seleniumPort = opts.seleniumPort || 4444;
```
You can pass the default value as the third argument or fourth argument(if collector function is present).

edemaine replied:
Indeed, you're right.  Thanks!  That wasn't obvious to me from the documentation...

edemaine replied:
There's actually a bug in the four-argument case: https://github.com/tj/commander.js/issues/686. But I was able to work around it.

ylemkimon commented on dockers/Screenshotter/screenshotter.js:
```diff
@@ -27,74 +27,32 @@ const newDir = path.normalize(
+    .option("--new",
+        "With `--verify`, generate new screenshots when match fails")
+    .option("--attempts <n>",
+        "Retry this many times before reporting failure (default 5)")
```
Shouldn't `--selenium-port`, `--katex-port`, and `--attempts` be collected as an integer(`parseInt`)?

ylemkimon commented:
@edemaine Most `babel` plugins depend on the working directory(`cwd`) being the root, where `package.json` and `.babelrc` resides. The working directory should be root and one can run screenshot tests using `npm run test:screenshots`.

ylemkimon commented on .circleci/config.yml:
```diff
@@ -21,7 +21,7 @@ defaults: &defaults
 
     - run:
         name: Verify screenshots and generate diffs and new screenshots
-        command: node dockers/Screenshotter/screenshotter.js --seleniumIP localhost -b $CIRCLE_JOB --verify --diff --new
```
How did this work?

edemaine replied:
We could probably add an alias for the old form, but given that the screenshotter is internal to developers, I doubt it will be too much trouble.

ylemkimon commented:
@edemaine Most `babel` plugins depend on the working directory(`cwd`) being the project root, where `package.json` and `.babelrc` resides. Maybe we should `pushd` the project root at the start of `dockers/Screenshotter/screenshotter.sh` and `popd` at the end.

edemaine commented:
@ylemkimon All comments implemented, and I also restore the `-m`/`--macro` behavior of accumulating multiple macros into an array, reading over the commander.js documentation more closely.

ylemkimon commented on cli.js:
```diff
@@ -1,103 +1,64 @@
+    .option("-m, --macro <def>",
+        "Define custom macro of the form '\\foo:expansion' (use multiple -m " +
+        "arguments for multiple macros).",
+        (def, defs) => defs.push(def),
```
`Array.push` returns the new `length`. This should be
```js
(def, defs) => {
    defs.push(def);
    return defs;
}
```

ylemkimon commented on cli.js:
```diff
@@ -1,103 +1,64 @@
+    console.log("KaTeX could not import, likely because dist/katex.js is missing.");
+    console.log("Please run 'npm install' and 'npm run dist' before running");
+    console.log("cli.js from the KaTeX repository.");
+    console.log();
```
I think `console.error` is more appropriate, as it print to `stderr`.

ylemkimon commented on package.json:
```diff
@@ -83,7 +83,7 @@
   },
   "dependencies": {
-    "nomnom": "^1.8.1"
+    "commander": "^2.16.0"
```
The file permissions of `package.json` and `package-lock.json` have been changed to 755.

edemaine commented:
Thanks. All fixed now.  (I thought I'd tested the macro arguments, but I guess not since I removed the braces and thereby changed the return value from `null`.  Tested again.)

I also changed the build instructions from `npm run dist` to `npm run build` given that #1500 is about to merge.

ylemkimon commented on cli.js:
```diff
@@ -1,103 +1,66 @@
+try {
+    katex = require("./");
+} catch (e) {
+    console.error("KaTeX could not import, likely because dist/katex.js is missing.");
```
>  13:1  error  Line 13 exceeds the maximum line length of 84  max-len

edemaine commented:
@ylemkimon Thanks for the reviews!
